File: Chapter04/FS/BlackCat/BlackCat/BlackCat/BlackCatPage.fs
namespace BlackCat

open System
open System.IO
open System.Reflection
open Xamarin.Forms

type BlackCatPage() = 
    inherit ContentPage(BackgroundColor = Color.White,
                        Padding = match Device.RuntimePlatform with 
                                        | Device.iOS -> Thickness(0.0, 20.0, 0.0, 0.0)
                                        | _ -> Thickness())

    // Create the two StackLayout's.
    let mainStack = StackLayout()
    let textStack = StackLayout(Padding = new Thickness(3.0),
                                Spacing = 10.0)

    // Get access to the text resource.
    let assembly = base.GetType().GetTypeInfo().Assembly
    let resource = "TheBlackCat.txt"

    // Get all the lines of the file.
    let lines = seq { use stream = assembly.GetManifestResourceStream(resource)
                      use reader = new StreamReader(stream)
                      while not reader.EndOfStream do
                          yield reader.ReadLine() }

    // Title line goes in mainStack.
    do mainStack.Children.Add(Label(Text = Seq.head lines,
                                    TextColor = Color.Black,    
                                    HorizontalOptions = LayoutOptions.Center,
                                    FontSize = Device.GetNamedSize(NamedSize.Medium, typeof<Label>),
                                    FontAttributes = FontAttributes.Bold))

    // Remaining lines go in textStack.
    do Seq.skip 1 lines
        |> Seq.iter (fun line -> textStack.Children.Add(Label(Text = line,
                                                              TextColor = Color.Black)))

    // Put ScrollView with textStack in mainstack.
    do mainStack.Children.Add(ScrollView(Content = textStack,
                                         VerticalOptions = LayoutOptions.FillAndExpand,
                                         Padding = Thickness(5.0, 0.0)))

    // Set mainStack to page content.
    do base.Content <- mainStack

File: Chapter04/BlackCat/BlackCat/BlackCat/BlackCatPage.cs
using System;
using System.IO;
using System.Reflection;
using Xamarin.Forms;

namespace BlackCat
{
    class BlackCatPage : ContentPage
    {
        public BlackCatPage()
        {
            StackLayout mainStack = new StackLayout();
            StackLayout textStack = new StackLayout
            {
                Padding = new Thickness(5),
                Spacing = 10
            };

            // Get access to the text resource.
            Assembly assembly = GetType().GetTypeInfo().Assembly;
            string resource = "BlackCat.Texts.TheBlackCat.txt";

            using (Stream stream = assembly.GetManifestResourceStream (resource)) 
            {
                using (StreamReader reader = new StreamReader (stream)) 
                {
                    bool gotTitle = false;
                    string line;

                    // Read in a line (which is actually a paragraph).
                    while (null != (line = reader.ReadLine())) 
                    {
                        Label label = new Label 
                        {
                            Text = line,

                            // Black text for ebooks!
                            TextColor = Color.Black
                        };

                        if (!gotTitle)
                        {
                            // Add first label (the title) to mainStack.
                            label.HorizontalOptions = LayoutOptions.Center;
                            label.FontSize = Device.GetNamedSize(NamedSize.Medium, label);
                            label.FontAttributes = FontAttributes.Bold;
                            mainStack.Children.Add(label);
                            gotTitle = true;
                        }
                        else
                        {
                            // Add subsequent labels to textStack.
                            textStack.Children.Add(label);
                        }
                    }
                }
            }

            // Put the textStack in a ScrollView with FillAndExpand.
            ScrollView scrollView = new ScrollView
            {
                Content = textStack,
                VerticalOptions = LayoutOptions.FillAndExpand,
                Padding = new Thickness(5, 0),
            };

            // Add the ScrollView as a second child of mainStack.
            mainStack.Children.Add(scrollView);

            // Set page content to mainStack.
            Content = mainStack;

            // White background for ebooks!
            BackgroundColor = Color.White;
    
            // Add some iOS padding for the page
            Padding = new Thickness (0, Device.RuntimePlatform == Device.iOS ? 20 : 0, 0, 0);
        }
    }
}

